File: agent_c/core/skills/run_command.py
from typing import Annotated
import subprocess
import asyncio
import platform
from .get_terminal_output import get_terminal_output  # Absolute import

async def run_command(terminal_id: Annotated[str, "Terminal ID"], command: Annotated[str, "Command to execute"]) -> Annotated[str, "Command Output"]:
    """
    Executes a shell command in a specified terminal and returns the output.

    Parameters:
    terminal_id (str): Identifier for the terminal in which the command should be run.
    command (str): The shell command to execute.

    Returns:
    str: The output from the terminal after running the command.
    """
    os_type = platform.system()
    try:
        # Run the command in the actual terminal
        if os_type == "Darwin":  # macOS
            run_script = f'''
            tell application "Terminal"
                do script "{command}" in window id {terminal_id}
            end tell
            '''
            subprocess.run(['osascript', '-e', run_script], capture_output=True, text=True)

        elif os_type == "Linux":  # Linux
            subprocess.run(['gnome-terminal', '--', 'bash', '-c', command])

        elif os_type == "Windows":  # Windows
            run_script = f'Start-Process powershell -ArgumentList "{command}" -NoNewWindow -PassThru'
            subprocess.run(['powershell', '-Command', run_script])

        # Custom message for 'cd' command
        if command.startswith("cd "):
            directory = command[3:].strip()
            # Actual terminal will execute the cd command without any output
            output = f"Navigated successfully to {directory}"
        else:
            # Capture the output using get_terminal_output
            output = await get_terminal_output(command)
            
        return output
    except Exception as e:
        return str(e)

if __name__ == "__main__":
    async def main():
        # Open a shell to get the terminal ID
        from agent_c.core.skills.open_shell import open_shell
        os_type = platform.system()
        terminal_id = open_shell(os_type)
        print(f"Opened terminal with ID: {terminal_id}")

        # List of commands to execute
        commands = ["cd ~/Desktop", "ls -la"]  # Example commands

        # Run each command and print the output
        for command in commands:
            output = await run_command(terminal_id, command)
            print("\n")
            print(f"Captured output for '{command}':\n{output}")

    asyncio.run(main())

File: agent_c/core/skills/run_command.cpp
#include <iostream>
#include <cstdio>
#include <cstdlib>
#include <string>
#include <stdexcept>
#include <array>

std::string exec(const char* cmd) {
    std::array<char, 128> buffer;
    std::string result;
    FILE* pipe = popen(cmd, "r");
    if (!pipe) {
        throw std::runtime_error("popen() failed!");
    }
    while (fgets(buffer.data(), buffer.size(), pipe) != nullptr) {
        result += buffer.data();
    }
    auto rc = pclose(pipe);
    if (rc == -1) {
        throw std::runtime_error("pclose() failed!");
    }
    return result;
}

int main(int argc, char *argv[]) {
    if (argc < 3) {
        std::cerr << "Usage: " << argv[0] << " <terminal_id> <command>" << std::endl;
        return 1;
    }

    std::string terminal_id = argv[1];
    std::string command = argv[2];

    // Construct the full command to run in the specified terminal
    std::string full_command = "osascript -e 'tell application \"Terminal\" to do script \"" + command + "\" in window id " + terminal_id + "'";

    // Execute the command and capture the output
    std::string output = exec(full_command.c_str());

    // Print the output
    std::cout << output << std::endl;

    return 0;
}
